diff -r -c orig/binutils-2.7/install.sh binutils-2.7/install.sh
*** orig/binutils-2.7/install.sh	Thu Jul 04 16:22:00 1996
--- binutils-2.7/install.sh	Mon Jun 23 09:34:50 1997
***************
*** 49,54 ****
--- 49,55 ----
  src=""
  dst=""
  dir_arg=""
+ exe_flag="true"
  
  while [ x"$1" != x ]; do
      case $1 in
***************
*** 61,66 ****
--- 62,68 ----
  	    continue;;
  
  	-m) chmodcmd="$chmodprog $2"
+ 	    exe_flag=""
  	    shift
  	    shift
  	    continue;;
***************
*** 147,152 ****
--- 149,164 ----
  		dst="$dst"/`basename $src`
  	else
  		true
+ 	fi
+ 
+ 	if [ x$exe_flag != x ]
+ 	then
+ 		dst=$dst.exe
+ 	fi
+ 
+ 	if [ -f $src.exe ]
+ 	then
+ 		src=$src.exe
  	fi
  fi
  
