diff -urN orig/binutils-2.7/ld/ldfile.c binutils-2.7/ld/ldfile.c
--- orig/binutils-2.7/ld/ldfile.c	Thu Jul  4 12:24:31 1996
+++ binutils-2.7/ld/ldfile.c	Fri Jun 13 17:19:02 1997
@@ -195,6 +195,11 @@
 	   arch != (search_arch_type *) NULL;
 	   arch = arch->next)
 	{
+
+          if (config.dynamic_link) {
+	    if (ldfile_open_file_search (arch->name, entry, "lib", ".sa"))
+	      return;
+	  }
 	  if (ldfile_open_file_search (arch->name, entry, "lib", ".a"))
 	    return;
 #ifdef VMS
