diff -urN orig-binutils-2.11.2/bfd/config.bfd binutils-2.11.2/bfd/config.bfd
--- orig-binutils-2.11.2/bfd/config.bfd	Thu Jun 14 11:16:59 2001
+++ binutils-2.11.2/bfd/config.bfd	Wed Jan  2 14:05:07 2002
@@ -161,6 +161,10 @@
     targ_defvec=bfd_elf32_littlearm_oabi_vec
     targ_selvecs=bfd_elf32_bigarm_oabi_vec
     ;;
+  arm-*-palmos*)
+    targ_defvec=bfd_elf32_littlearm_vec
+    targ_selvecs=bfd_elf32_bigarm_vec
+    ;;
 
   thumb-*-coff)
     targ_defvec=armcoff_little_vec
@@ -566,6 +570,9 @@
     targ_defvec=bfd_elf32_m68k_vec
     targ_selvecs=ieee_vec
     targ_underscore=yes
+    ;;
+  m68*-*-palmos*)
+    targ_defvec=m68kcoff_vec
     ;;
 
   m88*-harris-cxux* | m88*-*-dgux* | m88*-*-sysv4*)
diff -urN orig-binutils-2.11.2/config.sub binutils-2.11.2/config.sub
--- orig-binutils-2.11.2/config.sub	Thu Mar 22 11:12:23 2001
+++ binutils-2.11.2/config.sub	Wed Jan  2 14:05:07 2002
@@ -255,6 +255,10 @@
 		echo Invalid configuration \`$1\': machine \`$basic_machine\' not recognized 1>&2
 		exit 1
 		;;
+	m68k-palmos)
+		basic_machine=m68k-unknown
+		os=-palmos
+		;;
 	# Recognize the basic CPU types with company name.
 	# FIXME: clean up the formatting here.
 	vax-* | tahoe-* | i[234567]86-* | i860-* | ia64-* | m32r-* | m68k-* | m68000-* \
@@ -1017,6 +1021,7 @@
 	      | -mingw32* | -linux-gnu* | -uxpv* | -beos* | -mpeix* | -udk* \
 	      | -interix* | -uwin* | -rhapsody* | -darwin* | -opened* \
 	      | -openstep* | -oskit* | -conix* | -pw32* | -nonstopux* \
+	      | -palmos* \
 	      | -storm-chaos* | -tops10* | -tenex* | -tops20* | -its* | -os2*)
 	# Remember, each alternative MUST END IN *, to match a version number.
 		;;
diff -urN orig-binutils-2.11.2/gas/configure.in binutils-2.11.2/gas/configure.in
--- orig-binutils-2.11.2/gas/configure.in	Mon Jun 11 10:04:44 2001
+++ binutils-2.11.2/gas/configure.in	Wed Jan  2 14:05:07 2002
@@ -196,6 +196,7 @@
       arm-*-pe | thumb-*-pe)            fmt=coff em=pe ;;
       arm-*-riscix*)	                fmt=aout em=riscix ;;
       arm-*-vxworks)	                fmt=coff ;;
+      arm-*-palmos*)	                fmt=elf ;;
 
       avr-*-*)		    fmt=elf bfd_gas=yes ;;
 
@@ -321,6 +322,7 @@
       m68k-*-openbsd*)      fmt=aout em=nbsd bfd_gas=yes ;;
       m68k-apple-aux*)      fmt=coff em=aux ;;
       m68k-*-psos*)         fmt=elf em=psos;;
+      m68k-*-palmos*)       fmt=coff ;;
 
       m88k-motorola-sysv3*) fmt=coff em=delt88 ;;
       m88k-*-coff*)         fmt=coff ;;
diff -urN orig-binutils-2.11.2/gas/configure binutils-2.11.2/gas/configure
--- orig-binutils-2.11.2/gas/configure	Mon Jun 11 10:04:38 2001
+++ binutils-2.11.2/gas/configure	Wed Jan  2 14:05:07 2002
@@ -2312,6 +2312,7 @@
       arm-*-pe | thumb-*-pe)            fmt=coff em=pe ;;
       arm-*-riscix*)	                fmt=aout em=riscix ;;
       arm-*-vxworks)	                fmt=coff ;;
+      arm-*-palmos*)	                fmt=elf ;;
 
       avr-*-*)		    fmt=elf bfd_gas=yes ;;
 
@@ -2440,6 +2441,7 @@
       m68k-*-openbsd*)      fmt=aout em=nbsd bfd_gas=yes ;;
       m68k-apple-aux*)      fmt=coff em=aux ;;
       m68k-*-psos*)         fmt=elf em=psos;;
+      m68k-*-palmos*)       fmt=coff ;;
 
       m88k-motorola-sysv3*) fmt=coff em=delt88 ;;
       m88k-*-coff*)         fmt=coff ;;
diff -urN orig-binutils-2.11.2/ld/configure.tgt binutils-2.11.2/ld/configure.tgt
--- orig-binutils-2.11.2/ld/configure.tgt	Mon Jun 11 10:05:08 2001
+++ binutils-2.11.2/ld/configure.tgt	Wed Jan  2 14:05:07 2002
@@ -191,6 +191,7 @@
 arm*-*-linux-gnu*)	targ_emul=armelf_linux; targ_extra_emuls=armelf ;;
 arm*-*-uclinux*)	targ_emul=armelf_linux; targ_extra_emuls=armelf ;;
 arm*-*-conix*)		targ_emul=armelf ;;
+arm*-*-palmos*)		targ_emul=armelf ;;
 thumb-*-linux-gnu* | thumb-*-uclinux*)	targ_emul=armelf_linux; targ_extra_emuls=armelf ;;
 strongarm-*-coff)	targ_emul=armcoff ;;
 strongarm-*-elf)	targ_emul=armelf ;;
@@ -242,6 +243,7 @@
 m68*-*-netbsd*)		targ_emul=m68knbsd ;;
 m68*-*-psos*)		targ_emul=m68kpsos ;;
 m68*-*-rtems*)		targ_emul=m68kcoff ;;
+m68*-*-palmos*)		targ_emul=m68kcoff ;;
 hppa*64*-*)		targ_emul=elf64hppa ;;
 hppa*-*-linux-gnu*)	targ_emul=hppalinux ;;
 hppa*-*-*elf*)		targ_emul=hppaelf ;;
