diff -ur orig/prc-tools-2.0/configure prc-tools-2.0/configure
--- orig/prc-tools-2.0/configure	Sun Jan 30 18:10:26 2000
+++ prc-tools-2.0/configure	Sun Jan 30 18:23:53 2000
@@ -1093,6 +1093,10 @@
 	    grep pwplus $ac_dir/$ac_prog >/dev/null 2>&1; then
 	    # program-specific install script used by HP pwplus--don't use.
 	    :
+	  elif test $ac_prog = install &&
+	    grep cygwin1\.dll $ac_dir/$ac_prog.exe >/dev/null 2>&1; then
+	    # Cygwin install.exe sometimes can't create an existing directory.
+	    :
 	  else
 	    ac_cv_path_install="$ac_dir/$ac_prog -c"
 	    break 2
diff -ur orig/gcc-2.95.2/texinfo/makeinfo/makeinfo.c gcc-2.95.2/texinfo/makeinfo/makeinfo.c
--- orig/gcc-2.95.2/texinfo/makeinfo/makeinfo.c	Fri Jan 21 18:53:09 2000
+++ gcc-2.95.2/texinfo/makeinfo/makeinfo.c	Sat Jan 22 14:23:28 2000
@@ -933,6 +933,39 @@
      int argc;
      char **argv;
 {
+#if 1
+  /* JWM: Makeinfo simply doesn't seem to work on cygwin.  So we just
+     create a dummy output file to keep the Makefile happy.  We'll build
+     html docs instead of info on cygwin anyway.  */
+  char *outfname = NULL;
+  int i;
+
+  for (i = 1; i < argc; i++)
+    if (strncmp (argv[i], "-o", 2) == 0) {
+      outfname = &argv[i][2];
+      if (*outfname == '\0')  outfname = argv[i+1];
+      }
+
+  if (!outfname && strstr (argv[argc-1], ".texi")) {
+    static char buf[FILENAME_MAX];
+    char *slash = strrchr (argv[argc-1], '/');
+    strcpy (buf, (slash)? slash+1 : argv[argc-1]);
+    strcpy (strstr (buf, ".texi"), ".info");
+    outfname = buf;
+    }
+
+  if (outfname) {
+    FILE *f = fopen (outfname, "w");
+    fprintf (f, "Makeinfo not available.\n");
+    fclose (f);
+    }
+  else
+    fprintf (stderr, "%s: can't find an output filename\n", argv[0]);
+
+  return (outfname)? EXIT_SUCCESS : EXIT_FAILURE;
+
+#else
+
   extern int errors_printed;
   char *filename_part ();
   int c, ind;
@@ -1144,6 +1177,7 @@
     return (SYNTAX);
   else
     return (NO_ERROR);
+#endif
 }
 
 /* Display the version info of this invocation of Makeinfo. */
--- orig/binutils-2.9.1/ld/Makefile.in	Thu Jan 27 16:15:27 2000
+++ binutils-2.9.1/ld/Makefile.in	Thu Jan 27 16:16:08 2000
@@ -1136,9 +1136,7 @@
 # because almost all configs use "gen" version of manual.
 #  Set DOCVER above to change.
 configdoc.texi:	${DOCVER}-doc.texi
-	ln -s ${srcdir}/${DOCVER}-doc.texi ./configdoc.texi >/dev/null 2>&1 \
-	  || ln ${srcdir}/${DOCVER}-doc.texi ./configdoc.texi >/dev/null 2>&1 \
-	  || cp ${srcdir}/${DOCVER}-doc.texi ./configdoc.texi
+	  cp ${srcdir}/${DOCVER}-doc.texi ./configdoc.texi
 
 ld.info: $(srcdir)/ld.texinfo configdoc.texi
 	@rm -f $@ $@-[0-9] $@-[0-9][0-9]
diff -ur orig/binutils-2.9.1/gas/doc/Makefile.in binutils-2.9.1/gas/doc/Makefile.in
--- orig/binutils-2.9.1/gas/doc/Makefile.in	Thu Jan 27 16:16:19 2000
+++ binutils-2.9.1/gas/doc/Makefile.in	Thu Jan 27 16:16:38 2000
@@ -346,9 +346,7 @@
 
 asconfig.texi: $(CONFIG).texi
 	rm -f asconfig.texi
-	ln -s $(srcdir)/$(CONFIG).texi ./asconfig.texi >/dev/null 2>&1 \
-	  || ln $(srcdir)/$(CONFIG).texi ./asconfig.texi >/dev/null 2>&1 \
-	  || cp $(srcdir)/$(CONFIG).texi ./asconfig.texi
+	  cp $(srcdir)/$(CONFIG).texi ./asconfig.texi
 
 as.info: $(srcdir)/as.texinfo asconfig.texi $(CPU_DOCS)
 as.dvi: $(srcdir)/as.texinfo asconfig.texi $(CPU_DOCS)
