diff -urN orig-make-3.77/Makefile.in make-3.77/Makefile.in
--- orig-make-3.77/Makefile.in	Wed Jul 29 19:38:45 1998
+++ make-3.77/Makefile.in	Mon Jan 31 03:53:40 2000
@@ -257,6 +257,13 @@
 make.info: make.texinfo
 make.dvi: make.texinfo
 
+MAKEHTML = texi2html
+MAKEHTMLFLAGS = -glossary
+html: make_toc.html
+
+make_toc.html: $(srcdir)/make.texinfo
+	-@rm -f make[_.]*html
+	$(MAKEHTML) $(MAKEHTMLFLAGS) -I $(srcdir) $(srcdir)/make.texinfo
 
 DVIPS = dvips
 
diff -urN orig-make-3.77/make.texinfo make-3.77/make.texinfo
--- orig-make-3.77/make.texinfo	Mon Jul 27 21:15:17 1998
+++ make-3.77/make.texinfo	Mon Jan 31 04:23:55 2000
@@ -354,6 +354,9 @@
 @ifinfo
 @heading Preparing and Running Make
 @end ifinfo
+@iftex
+@section Preparing and Running Make
+@end iftex
 
 To prepare to use @code{make}, you must write a file called
 the @dfn{makefile} that describes the relationships among files
@@ -1340,7 +1343,6 @@
                                  dependencies from the source files themselves.
 @end menu
 
-@ifinfo
 @node Rule Example, Rule Syntax,  , Rules
 @section Rule Example
 
@@ -1369,7 +1371,6 @@
 that @file{foo.c} includes it, and that that is why @file{defs.h} was
 added to the dependencies.
 @end itemize
-@end ifinfo
 
 @node Rule Syntax, Wildcards, Rule Example, Rules
 @section Rule Syntax
